### ISSUE TITLE
Inherit inner class annotated with @Group when extending test container class

### ISSUE THREAD

**Issue by DirkToewe** [jlink/jqwik#179]
## Testing Problem

`test2()` in the following code cannot be executed, because `TestGroup` cannot be instantiated.

```java
abstract class AC {
  @Group public class TestGroup
  {
    @Property void test2( @ForAll int i ) {
      assert i >= Integer.MIN_VALUE;
    }
  }
}
@Group
public class MWE_Test extends AC {
  @Property void test1( @ForAll int i ) {
    assert i <= Integer.MAX_VALUE;
  }
}
```

## Suggested Solution

After taking a quick glance at the source code, my poorly educated guess is that `TestGroup` is not recognized to be a member class by the method `JqwikReflectionSupport.isInnerClass()`. Maybe `isInnerClass()` could look through the constructors as well, something along the lines of:

```java
public boolean isInnerClass( Class<?> clazz ) {
  var decl = clazz.getDeclaringClass();
  if( decl == null ) return false;

  if( ! Modifier.isStatic(clazz.getModifiers()) )
    return true;

  var constructorFound = false;

  for( var con: clazz.getConstructors() )
  {
    var par = con.getParameterTypes();
    if( par.length == 0 ) return false; // <- is statically instantiable
    if( par.length == 1 && par[0].isAssignableFrom(decl) )
      constructorFound = true;
  }

  return constructorFound;
}
```

## Discussion

Being able to extend abstract test (template) classes involving Groups would greatly improve reusability of tests and reduces the amount of code written.

**Comment by jlink** [comment]
@DirkToewe This would probably be doable. Why I'm a bit hesitant: Is it the expected thing from a Java developer's point of view that extending a class will also "inherit" the extended class' inner classes. How does Jupiter handle this same situation?

**Comment by DirkToewe** [comment]
The following junit-juper example runs just fine including the `@Nested` test:
```java
abstract class AClassTest {
  @Nested class BClassTest {
    @Test void testicle() {
      assertThrows(Error.class, () -> { throw new Error(); });
    }
  }
}
public class NestTest extends AClassTest {
  @Test void test() {
    assertThrows(Error.class, () -> { throw new Error(); });
  }
}
```

How exactly this is implemented in junit-jupiter, is beyond my understanding. I am however not sure what alternative there is to my solution above is. Ignore the `@Group`, but when would that be useful?

**Comment by jlink** [comment]
Jupiter doing it that way is a strong argument for me. I'll check how they do it.
So this gets in my list of features to prioritize. @DirkToewe, Thanks for bringing it up!

**Comment by jlink** [comment]
Status now #InProgress

**Comment by jlink** [comment]
@DirkToewe I just ran your original example and I get the following error:

```
net.jqwik.api.JqwikException: Cannot create instance of class 'class examples.AC$TestGroup'
```

Is that the same behaviour you observed? I got the impression that you said jqwik would not even detect TestGroup as an inner test of MWE_Test.

**Comment by jlink** [comment]
Should be fixed in https://github.com/jlink/jqwik/commit/c17f16e6bc5f317d40add1ada064897b365114b6.

Available in 1.5.2-SNAPSHOT.

@DirkToewe Maybe you find the time to check if your problem is solved.

**Comment by DirkToewe** [comment]
It is working as desired now. This will save me a lot of lines of code in the future. Thank You :)
